File: csharp/WebSocketService.cs
using System;
using System.Net.Http;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

public class WebSocketService
{
	private readonly string accountId;
	private readonly string clientId;
	private readonly string clientSecret;
	private readonly string url;
	private readonly string oauthUrl = "https://zoom.us/oauth/";
	private ClientWebSocket ws;
	private CancellationTokenSource heartbeatTokenSource;

	public WebSocketService(string accountId, string clientId, string clientSecret, string url)
	{
		this.accountId = accountId;
		this.clientId = clientId;
		this.clientSecret = clientSecret;
		this.url = url;
	}

	private async Task<string> GetAccessTokenAsync()
	{
		try
		{
			var authToken = Convert.ToBase64String(Encoding.UTF8.GetBytes($"{clientId}:{clientSecret}"));
			using var client = new HttpClient();
			client.DefaultRequestHeaders.Add("Authorization", $"Basic {authToken}");

			var response = await client.PostAsync(
				$"{oauthUrl}token?grant_type=account_credentials&account_id={accountId}",
				null);

			response.EnsureSuccessStatusCode();
			var content = await response.Content.ReadAsStringAsync();
			var json = JsonSerializer.Deserialize<JsonElement>(content);

			return json.GetProperty("access_token").GetString();
		}
		catch (Exception ex)
		{
			Console.WriteLine($"Error retrieving access token: {ex.Message}");
			throw;
		}
	}

	private async Task SendHeartbeatAsync()
	{
		while (!heartbeatTokenSource.Token.IsCancellationRequested)
		{
			var heartbeatMessage = JsonSerializer.Serialize(new { module = "heartbeat" });
			var buffer = Encoding.UTF8.GetBytes(heartbeatMessage);

			await ws.SendAsync(new ArraySegment<byte>(buffer), WebSocketMessageType.Text, true, CancellationToken.None);

			await Task.Delay(TimeSpan.FromSeconds(30), heartbeatTokenSource.Token);
		}
	}

	private void OnMessageReceived(string message)
	{
		try
		{
			var dataObj = JsonSerializer.Deserialize<JsonElement>(message);

			if (dataObj.TryGetProperty("module", out var module) && module.GetString() == "message")
			{
				var contentString = dataObj.GetProperty("content").GetString();
				var contentObj = JsonSerializer.Deserialize<JsonElement>(contentString);

				Console.WriteLine(JsonSerializer.Serialize(
					contentObj,
					new JsonSerializerOptions { WriteIndented = true }
				));

				if (contentObj.GetProperty("event").GetString() == "user.created")
				{
					NewUserCreatedHandler();
				}
			}
		}
		catch (Exception ex)
		{
			Console.WriteLine($"Invalid JSON received: {ex.Message}");
		}
	}

	private void NewUserCreatedHandler()
	{
		Console.WriteLine("\n\nA new user was created");
		Console.WriteLine("Do some processing\n\n");
	}

	public async Task ConnectAsync()
	{
		try
		{
			var accessToken = await GetAccessTokenAsync();

			ws = new ClientWebSocket();
			await ws.ConnectAsync(new Uri($"{url}&access_token={accessToken}"), CancellationToken.None);

			Console.WriteLine("Connected");

			heartbeatTokenSource = new CancellationTokenSource();
			_ = Task.Run(SendHeartbeatAsync);

			var buffer = new byte[1024 * 4];

			while (ws.State == WebSocketState.Open)
			{
				var result = await ws.ReceiveAsync(new ArraySegment<byte>(buffer), CancellationToken.None);

				if (result.MessageType == WebSocketMessageType.Text)
				{
					var message = Encoding.UTF8.GetString(buffer, 0, result.Count);
					OnMessageReceived(message);
				}
				else if (result.MessageType == WebSocketMessageType.Close)
				{
					Console.WriteLine("Connection closed");
					await ws.CloseAsync(WebSocketCloseStatus.NormalClosure, string.Empty, CancellationToken.None);
				}
			}
		}
		catch (Exception ex)
		{
			Console.WriteLine($"Error: {ex.Message}");
		}
	}
}

File: rust/src/websocket_service.rs
use base64::prelude::*;
use futures_util::SinkExt;
use futures_util::StreamExt;
use reqwest::Client;
use serde_json::Value;
use std::time::Duration;
use tokio::sync::mpsc;
use tokio::time;
use tokio_tungstenite::connect_async;
use tungstenite::protocol::Message;

pub struct WebSocketService {
    account_id: String,
    client_id: String,
    client_secret: String,
    url: String,
    oauth_url: String,
}

impl WebSocketService {
    pub fn new(account_id: String, client_id: String, client_secret: String, url: String) -> Self {
        Self {
            account_id,
            client_id,
            client_secret,
            url,
            oauth_url: "https://zoom.us/oauth/".to_string(),
        }
    }

    pub async fn get_access_token(&self) -> Option<String> {
        let oauth_token =
            BASE64_STANDARD.encode(format!("{}:{}", self.client_id, self.client_secret));
        let client = Client::new();

        let response = client
            .post(format!(
                "{}token?grant_type=account_credentials&account_id={}",
                self.oauth_url, self.account_id
            ))
            .header("Authorization", format!("Basic {}", oauth_token))
            .send()
            .await;

        match response {
            Ok(resp) => {
                if let Ok(json) = resp.json::<Value>().await {
                    return json["access_token"].as_str().map(|s| s.to_string());
                }
            }
            Err(e) => {
                eprintln!("Error fetching access token: {}", e);
            }
        }

        None
    }

    pub async fn connect(&self) {
        if let Some(access_token) = self.get_access_token().await {
            let ws_url = format!("{}&access_token={}", self.url, access_token);
            match connect_async(&ws_url).await {
                Ok((websocket, _)) => {
                    println!("Connected to WebSocket.");

                    // Split the websocket into sender and receiver
                    let (mut ws_sender, mut ws_receiver) = websocket.split();

                    let (tx, mut rx) = mpsc::channel::<Message>(32);
                    let tx_clone = tx.clone();

                    // Spawn task for sending messages
                    tokio::spawn(async move {
                        while let Some(msg) = rx.recv().await {
                            ws_sender.send(msg).await.unwrap();
                        }
                    });

                    self.send_heartbeat(tx_clone);

                    // Use receiver in the main loop
                    while let Some(msg) = ws_receiver.next().await {
                        match msg {
                            Ok(Message::Text(text)) => self.message_handler(&text).await,
                            Ok(_) => {}
                            Err(e) => {
                                eprintln!("WebSocket error: {}", e);
                                break;
                            }
                        }
                    }

                    println!("Connection closed.");
                }
                Err(e) => {
                    eprintln!("Failed to connect to WebSocket: {}", e);
                }
            }
        } else {
            eprintln!("Unable to get access token.");
        }
    }

    fn send_heartbeat(&self, tx: mpsc::Sender<Message>) {
        tokio::spawn(async move {
            loop {
                time::sleep(Duration::from_secs(30)).await;
                if tx
                    .send(Message::Text("{\"module\": \"heartbeat\"}".to_string()))
                    .await
                    .is_err()
                {
                    break;
                }
            }
        });
    }

    pub async fn message_handler(&self, message: &str) {
        match serde_json::from_str::<Value>(message) {
            Ok(data) => {
                if let Some(module) = data["module"].as_str() {
                    if module == "message" {
                        if let Ok(pretty_json) = serde_json::to_string_pretty(&data["content"]) {
                            println!("Received content:\n{}", pretty_json);
                        }
                        if let Some(content) = data["content"].as_object() {
                            if let Some(event) = content["event"].as_str() {
                                if event == "user.created" {
                                    self.new_user_created_handler().await;
                                }
                            }
                        }
                    }
                }
            }
            Err(_) => {
                eprintln!("Invalid JSON received.");
            }
        }
    }

    async fn new_user_created_handler(&self) {
        println!("\n\nA new user was created.");
        println!("Perform custom processing here\n\n");
    }
}
